.github/workflows/swift_matrix.yml

```diff
@@ -108,6 +112,16 @@ jobs:
       if: ${{ matrix.swift.enabled }}
     # https://github.com/actions/checkout/issues/766
       run: git config --global --add safe.directory ${GITHUB_WORKSPACE}
+
+    - name: Restore .build
+      if: ${{ matrix.swift.enabled && (inputs.cache_key_prefix != '') }}
+      id: "restore-cache"
+      uses: actions/cache/restore@v4
+      with:
+        path: .build
+        key: "${{ inputs.cache_key_prefix }}${{ github.event.repository.name }}-build-${{ matrix.swift.swift_version }}-${{ runner.os }}-${{ (github.event.pull_request.base.sha != null) && github.event.pull_request.base.sha || github.event.after }}"
```

Comment:
`${{ (github.event.pull_request.base.sha != null) && github.event.pull_request.base.sha || github.event.after }}`
is the main thing that can be different for a specific swift version.
It allows the cache to be updated after each push to the main branch, since it'll resolve to `github.event.after`.

It either resolves to the base branch's SHA in PR events, or to the new SHA of the branch in push events.
So each PR will be looking for their base branch's cached .build.
And on a push to the main branch, the cache is refreshed.
It will be suboptimal, but even the event isn't a push or PR, the CI should still use the cache. But maybe also waste some seconds on trying to re-save the cache.

One confusing behavior of the caches action is that in a PR, you cannot create a cache that the next commits in a PR can use. That's why the cache "key" doesn't try to do something like that.
You can only access the repo's primary-branch's caches.

FYI another gotcha of the caches action is that you cannot delete OR override value of a cache "key".